File: cmake/check_test_results.rb
# Souffle - A Datalog Compiler
# Copyright (c) 2021 The Souffle Developers. All rights reserved
# Licensed under the Universal Permissive License v 1.0 as shown at:
# - https://opensource.org/licenses/UPL
# - <souffle root>/licenses/SOUFFLE-UPL.txt


# Expects the current working directory to be a test results directory.


require 'fileutils'

$: << __dir__
require 'common'

args = ARGV.dup

raise "missing INPUT_DIR" if args.empty?
input_dir = args.shift

extra_data = nil
binary = nil

if !args.empty?
  arg = args.shift
  case arg
  when ""
    extra_data = nil
  when "sqlite3", "gzip", "json"
    extra_data = arg
    raise "Missing BINARY" if args.empty?
    binary = args.shift
  when "json"
    extra_data = arg
  else
    raise "Unknown processing type #{arg}"
  end
end

raise "Unexpected argument" unless args.empty?

case extra_data
when "gzip"
  extra_file_pattern = "*.gz.output"
when "sqlite3"
  extra_file_pattern = "*.sqlite.output"
when "json"
  extra_file_pattern = "*.json"
else
  extra_file_pattern = nil
end

if extra_file_pattern
  extra_files = Dir.glob(extra_file_pattern)

  extra_files.each do |file|
    case extra_data
    when "gzip"
      generated_file = File.basename(file, ".gz.output")
      `"#{binary}" -d -c "#{file}" > "#{generated_file}"`
    when "sqlite3"
      generated_file = "%s.csv" % [File.basename(file, ".sqlite.output")]
      `"#{binary}" -batch "#{file}" -init "#{input_dir}/#{file}.script" "" > "#{generated_file}"`
    when "json"
      generated_file = file
    end

    sort_file(generated_file)
    compare_sorted_file(generated_file)
  end
end

generated_csvs = Dir.glob("*.csv")

File.open("num.generated", "w") do |num_file|
  num_file.puts generated_csvs.size
end

compare_files("num.expected", "num.generated")

csvs = Dir.glob("*.csv.expected.sorted").map{|file| File.basename(file, ".expected.sorted")}

csvs.each do |csv_file|
  sort_file(csv_file)
  compare_sorted_file(csv_file)
end

File: cmake/check_test_results.py
# Souffle - A Datalog Compiler
# Copyright (c) 2022 The Souffle Developers. All rights reserved
# Licensed under the Universal Permissive License v 1.0 as shown at:
# - https://opensource.org/licenses/UPL
# - <souffle root>/licenses/SOUFFLE-UPL.txt

import os
import glob
import subprocess
from common import *

args = os.sys.argv
args.pop(0)

if len(args) == 0:
    raise RuntimeError("Missing INPUT_DIR")

input_dir = args.pop(0)
extra_data = None
binary = None

if len(args) > 0:
    arg = args.pop(0)
    if arg == "":
        extra_data = None
    elif arg == "sqlite3" or arg == "gzip":
        extra_data = arg
        if len(args) == 0:
            raise RuntimeError("Missing BINARY")
        binary = args.pop(0)
    elif arg == "json":
        extra_data = arg
        if len(args) > 0:
            args.pop(0) # unused last argument
    else:
        raise RuntimeError("Unknown processing type {}".format(arg))

if len(args) != 0:
    raise RuntimeError("Unexpected argument")

if extra_data == "gzip":
    extra_file_pattern = "*.gz.output"
elif extra_data == "sqlite3":
    extra_file_pattern = "*.sqlite.output"
elif extra_data == "json":
    extra_file_pattern = "*.json"
else:
    extra_file_pattern = None

if extra_file_pattern:
    extra_files = glob.glob(extra_file_pattern)

    for file in extra_files:
        if extra_data == "gzip":
            generated_file = os.path.basename(file).rstrip(".gz.output")
            with open(generated_file, "w") as generated:
                subprocess.run([binary, "-d", "-c", file], check=True, stdout=generated)
        elif extra_data == "sqlite3":
            generated_file = "{}.csv".format(os.path.basename(file).rstrip(".sqlite.output"))
            script = os.path.join(input_dir, "{}.script".format(file))
            with open(generated_file, "w") as generated:
                subprocess.run([binary,"-batch","-init",script,file,""], check=True, stdout=generated)
        elif extra_data == "json":
            generated_file = file

        sort_file(generated_file)
        compare_sorted_file(generated_file)

generated_csvs = glob.glob("*.csv")

with open("num.generated", "w") as num_file:
    num_file.write("{}\n".format(len(generated_csvs)))

compare_files("num.expected", "num.generated")

csvs = list(map(
        lambda file: os.path.basename(file).rstrip(".expected.sorted"),
        glob.glob("*.csv.expected.sorted")))

for csv_file in csvs:
    sort_file(csv_file)
    compare_sorted_file(csv_file)

